File: pysrc/libs/term/parse_termdata.py
import logging
import re
import sys


class Cap:
    def __init__(self):
        self.flags = {}
        self.cmds = {}
        self.control_data_start_state = ControlDataState()


def parse_cap(cap_str):
    cap = Cap()

    for field in cap_str.split(':'):
        if len(field) == 0:
            continue

        if field.find('=') > 0:
            cap.cmds.update(parse_str_cap(field, cap.control_data_start_state))
        elif field.find('#') > 0:
            parts = field.split('#')
            cap.flags.update({parts[0]: int(parts[1])})
        else:
            cap.flags.update({field: 1})

    return cap


class ControlDataParserContext:
    def __init__(self):
        self.params = []

    def push_param(self, param):
        self.params.append(param)


class ControlDataState:
    def __init__(self):
        self.cap_name = {}
        self.next_states = {}
        self.digit_state = None
        self.any_state = None

    def reset(self):
        pass

    def add_state(self, c, state):
        if c in self.next_states:
            return self.next_states[c]

        self.next_states[c] = state
        return state

    def add_digit_state(self, state):
        if self.digit_state:
            return self.digit_state

        self.digit_state = state
        return state

    def add_any_state(self, state):
        if self.any_state:
            return self.any_state

        self.any_state = state
        return state

    def handle(self, context, c):
        if c in self.next_states:
            next_state = self.next_states[c]
            if next_state.digit_state:
                next_state.digit_state.reset()

            return next_state

        if self.digit_state:
            next_state = self.digit_state.handle(context, c)

            if next_state:
                return next_state

        if self.any_state:
            return self.any_state.handle(context, c)

        return None

    def get_cap(self, params):
        if len(params) == 0:
            if '' in self.cap_name:
                return self.cap_name['']

            if '***' in self.cap_name:
                return self.cap_name['***']

            return None

        str_match = ','.join([str(x) for x in params])

        if str_match in self.cap_name:
            return self.cap_name[str_match]

        for k in sorted(self.cap_name, key=lambda v: str(v.count('*')) + v):
            if k.find('*') < 0:
                if k == str_match:
                    return self.cap_name[k]
                else:
                    continue

            re_str = k.replace('***', '(.:)')
            re_str = re_str.replace(',**', '(,[0-9]+)?')
            re_str = re_str.replace('**', '([0-9]+)?')
            re_str = re_str.replace('*', '[0-9]+')
            re_str = re_str.replace('?', '*')
            re_str = re_str.replace(':', '*')

            if re.match(re_str, str_match):
                return self.cap_name[k]

        return None


class DigitState(ControlDataState):
    def __init__(self):
        ControlDataState.__init__(self)
        self.digit_base = 10
        self.digits = ['0', '1', '2', '3', '4', '5',
                       '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F']
        self.value = None

    def handle(self, context, cc):
        c = cc.upper()

        if c in self.digits[:self.digit_base]:
            self.value = self.value * self.digit_base + \
                self.digits.index(c) \
                if self.value else self.digits.index(c)

            return self
        else:
            if self.value is not None:
                context.push_param(self.value)
                self.value = None
            return ControlDataState.handle(self, context, cc)

    def reset(self):
        self.value = None


class AnyState(ControlDataState):
    def __init__(self):
        ControlDataState.__init__(self)

    def handle(self, context, cc):
        next_state = ControlDataState.handle(self, context, cc)

        if next_state:
            return next_state

        context.push_param(cc)

        return self


class CapStringValue:
    def __init__(self):
        self.padding = 0.0
        self.value = ''
        self.name = ''

    def __str__(self):
        return ','.join([self.name, str(self.padding), self.value])

    def __repr__(self):
        return self.__str__()


def parse_padding(value):
    padding = 0.0
    pos = 0

    if value[0].isdigit():
        padding_chars = []
        has_dot = False

        for pos in range(len(value)):
            if value[pos] == '*':
                if len(padding_chars) > 0:
                    padding_chars.append('*')
                    pos += 1
                    break
            elif value[pos] == '.':
                if has_dot:
                    break
                has_dot = True
                padding_chars.append('.')
            elif value[pos].isdigit():
                padding_chars.append(value[pos])
            else:
                break
        #end for

        try:
            padding = 0.0
            if padding_chars[-1] == '*':
                padding = float(''.join(padding_chars[:-1]))
            else:
                padding = float(''.join(padding_chars))
        except ValueError:
            pass

    return (pos, padding)


def build_parser_state_machine(cap_str_value, start_state):
    value = cap_str_value.value

    pos = 0

    cur_state = start_state
    repeat_state = None
    repeat_char = None
    is_repeat_state = False
    repeat_enter_state = None

    increase_param = False
    is_digit_state = False
    digit_base = 10
    params = []
    cap_value = []

    while pos <len(value):
        c = value[pos]

        if c == '\\':
            pos += 1

            if pos >= len(value):
                raise ValueError("Unterminaled str")

            c = value[pos]
            if c == 'E':
                c = chr(0x1B)
            elif c == '\\':
                c = '\\'
            elif c == '(':
                is_repeat_state = True
                repeat_enter_state = cur_state
                pos += 1
                continue
            elif c == ')':
                if not repeat_state or not repeat_char:
                    raise ValueError("Invalid repeat state:" + str(pos) + "," + value)

                cur_state.add_state(repeat_char, repeat_state)

                repeat_char = None
                repeat_state = None
                is_repeat_state = False
                pos += 1
                continue
            elif c.isdigit():
                v = 0
                while pos < len(value) and c.isdigit():
                    v = v * 8 + int(c)
                    pos += 1
                    if pos < len(value):
                        c = value[pos]

                if not c.isdigit():
                    pos -= 1

                c = chr(v)
            elif c == '.':
                cur_state = cur_state.add_any_state(AnyState())
                params.append('***')
                pos += 1
                continue
            else:
                raise ValueError("unknown escape string:" + c + "," + str(pos) + "," + value)
        elif c == '^':
            pos += 1

            if pos >= len(value):
                raise ValueError("Unterminaled str")

            c = chr(ord(value[pos]) - ord('A') + 1)
        elif c == '%':
            pos += 1

            if pos >= len(value):
                raise ValueError("Unterminaled str")

            c = value[pos]

            if c == '%':
                c = '%'
            elif c == 'i':
                increase_param = True
                pos += 1
                continue
            elif c == 'd':
                is_digit_state = True
                digit_base = 10
            elif c == 'X' or c == 'x':
                is_digit_state = True
                digit_base = 16
            else:
                raise ValueError('unknown format string:' + c + "," + str(pos) + "," + value)
        elif c.isdigit():
            v = 0
            while pos < len(value) and c.isdigit():
                v = v * 10 + int(c)
                pos += 1
                if pos < len(value):
                    c = value[pos]

            if not c.isdigit():
                #restore the last digit
                pos -= 1
                #include all digit char into cap_value
                c = str(v)

            #save the params
            params.append(str(v))

        #build state with c
        if is_digit_state:
            cur_state = cur_state.add_digit_state(DigitState())
            cur_state.digit_base = digit_base
            #save the params
            if is_repeat_state:
                params.append('**')
            else:
                params.append('*')
        elif c.isdigit():
            cur_state = cur_state.add_digit_state(DigitState())
            cap_value.append(c)
        else:
            cur_state = cur_state.add_state(c, ControlDataState())
            cap_value.append(c)

        if is_repeat_state and not repeat_state:
            repeat_state = cur_state
            repeat_char = c

        if not is_repeat_state and repeat_enter_state:
            old_cur_state = cur_state
            cur_state = repeat_enter_state.add_state(c, cur_state)

            if cur_state != old_cur_state:
                #merge the state
                logging.error('should put generic pattern before special pattern')
                sys.exit(1)
            repeat_enter_state = None

        is_digit_state = False

        pos += 1

    return (cur_state, params, increase_param, ''.join(cap_value))

def parse_str_cap(field, start_state):
    cap_str_value = CapStringValue()

    parts = field.split('=')

    cap_str_value.name = parts[0]
    value = cap_str_value.value = '='.join(parts[1:])

    #padding
    pos, cap_str_value.padding = parse_padding(value)

    #build the parser state machine
    value = cap_str_value.value = value[pos:]

    cap_state, params, increase_param, cap_str_value.cap_value = build_parser_state_machine(cap_str_value, start_state)

    cap_name_key = ','.join(params)

    if cap_name_key in cap_state.cap_name:
        e_name, e_inc_param = cap_state.cap_name[cap_name_key]

        if (e_name != cap_str_value.name) or (e_inc_param != increase_param):
            raise ValueError('same parameter for different cap name:[' + cap_name_key + '],' + cap_str_value.name)

    cap_state.cap_name[cap_name_key] = (cap_str_value.name, increase_param)

    return {parts[0]:cap_str_value}

File: src/caps/parse_termdata.cpp
#include "parse_termdata.h"

#include <sstream>

static
void parse_str_cap(const std::string & field,
                   CapPtr cap);

CapPtr parse_cap(const std::string & cap_str)
{
    CapPtr cap = std::make_shared<Cap>();

    auto cap_fields = tokenize(cap_str, ":");

    for(auto & field : cap_fields) {
        if (field.length() == 0)
            continue;

        if (field.find("=") != std::string::npos) {
            parse_str_cap(field,
                          cap);
        } else if (field.find("#") != std::string::npos) {
            auto parts = tokenize(field, "#");

            cap->flags.emplace(parts[0], atoi(parts[1].c_str()));
        } else {
            cap->flags.emplace(field, 1);
        }
    }

    return cap;
}

static
void parse_padding(std::string & value,
                   std::string::size_type & pos,
                   double & padding)
{
    pos = 0;
    padding = 0.0;

    if (!(value[0] >= '0' && value[0] <= '9')) {
        return;
    }

    std::string padding_chars;
    bool get_dot = false;

    for(;pos < value.length(); pos++) {
        if (value[pos] == '*') {
            pos++;
            break;
        } else if (value[pos] == '.') {
            if (get_dot)
                break;
            get_dot = true;
            padding_chars.append(".");
        } else if (value[pos] >= '0' && value[pos] <= '9') {
            padding_chars.append(value.substr(pos, 1));
        } else {
            break;
        }
    }

    padding = atof(padding_chars.c_str());
}

static
void build_parser_state_machine(CapStringValuePtr & cap_str_value,
                                ControlDataStatePtr & state)
{
    std::string value = cap_str_value->value;

    std::string::size_type pos = 0;
    ControlDataStatePtr cur_state = state;
    ControlDataStatePtr repeat_state {};
    char repeat_char = 0;
    bool is_repeat_state = false;
    ControlDataStatePtr repeat_enter_state {};

    bool increase_param = false;
    bool is_digit_state = false;
    int digit_base = 10;
    string_vector_t params;
    std::string cap_value;
    bool const_digit_state = false;
    std::string const_digit;

     while (pos < value.length()) {
        char c = value[pos];

        switch (c) {
        case '\\': {
            pos ++;

            if (pos >= value.length())
                throw "unterminated cap str value";

            c = value[pos];

            switch(c){
            case 'E':
                c = 0x1B;
                break;
            case '\\':
                c = '\\';
                break;
            case '(':
                is_repeat_state = true;
                repeat_enter_state = cur_state;
                pos++;
                continue;
            case ')': {
                if (!repeat_state || repeat_char == 0) {
                    std::stringstream ss;
                    ss << "invalid repeat state:"
                       << pos
                       << ","
                       << value;
                    throw ss.str();
                }

                cur_state->add_state(repeat_char, repeat_state);
                repeat_char = 0;
                repeat_state = nullptr;
                is_repeat_state = false;
                pos++;
                continue;
            }
                break;
            case '0':
            case '1':
            case '2':
            case '3':
            case '4':
            case '5':
            case '6':
            case '7':
            case '8':
            case '9': {
                int v = 0;
                while (pos < value.length() && c >= '0' && c <= '9') {
                    v = v * 8 + c - '0';
                    pos++;
                    if (pos < value.length())
                        c = value[pos];
                }

                if (!(c >= '0' && c <= '9'))
                    pos--;

                c = v & 0xFF;
            }
                break;
            default:{
                std::stringstream ss;
                ss << "invalid escape string:"
                   << pos
                   << ","
                   << value;
                throw ss.str();
            }
                break;
            }
        }
            break;
        case '^':
            pos++;
            if (pos >= value.length())
                throw "unterminated cap str value";
            c = value[pos] - 'A' + 1;
            break;
        case '%': {
            pos++;
            if (pos >= value.length())
                throw "unterminated cap str value";

            c = value[pos];
            switch(c){
            case '%':
                c = '%';
                break;
            case 's':
                cur_state = cur_state->add_any_state(std::make_shared<AnyState>());
                params.push_back("***");
                pos++;
                continue;
            case 'i':
                increase_param = true;
                pos++;
                continue;
            case 'd':
                is_digit_state = true;
                digit_base = 10;
                break;
            case 'X':
            case 'x':
                is_digit_state = true;
                digit_base = 16;
                break;
            default:{
                std::stringstream ss;
                ss << "invalid format string:"
                   << pos
                   << ","
                   << value;
                throw ss.str();
            }
                break;
            }
        }
            break;
        case '0':
        case '1':
        case '2':
        case '3':
        case '4':
        case '5':
        case '6':
        case '7':
        case '8':
        case '9': {
            int v = 0;
            while (pos < value.length() && c >= '0' && c <= '9') {
                v = v * 10 + c - '0';
                pos++;
                if (pos < value.length())
                    c = value[pos];
            }

            if (!(c >= '0' && c <= '9'))
                pos--;

            std::stringstream ss;
            ss << v;

            params.push_back(ss.str());
            const_digit_state = true;
            const_digit = ss.str();
        }
            break;
        default:
            break;
        }//switch

        if (is_digit_state) {
            DigitStatePtr digit_state = cur_state->add_digit_state(std::make_shared<DigitState>());
            digit_state->digit_base = digit_base;
            cur_state = digit_state;

            if (is_repeat_state)
                params.push_back("**");
            else
                params.push_back("*");
        } else if (const_digit_state) {
            DigitStatePtr digit_state =
                    cur_state->add_digit_state(std::make_shared<DigitState>());
            cur_state = digit_state;
            digit_state->digit_base = 10;
            cap_value.append(const_digit);
        } else {
            cur_state = cur_state->add_state(c, std::make_shared<ControlDataState>());
            cap_value.append(&c, 1);
        }

        if (is_repeat_state && !repeat_state) {
            repeat_state = cur_state;
            repeat_char = c;
        }

        if (!is_repeat_state && repeat_enter_state) {
            ControlDataStatePtr old_cur_state = cur_state;
            cur_state = repeat_enter_state->add_state(c, cur_state);

            if (cur_state != old_cur_state) {
                throw "should put generic pattern before special pattern";
            }

            repeat_enter_state = nullptr;
        }

        is_digit_state = false;
        const_digit_state = false;
        pos++;
    }//while

    cap_str_value->value = cap_value;

    std::string cap_name_key = join(params, ",");

    cur_state->update_cap_name(cap_name_key,
                               cap_str_value->name,
                               increase_param);
}

void parse_str_cap(const std::string & field,
                   CapPtr cap)
{
    CapStringValuePtr cap_str_value = std::make_shared<CapStringValue>();

    std::string::size_type pos = field.find("=");
    cap_str_value->name = field.substr(0, pos);
    std::string value = cap_str_value->value = field.substr(pos + 1);

    parse_padding(value, pos, cap_str_value->padding);
    value = cap_str_value->value = value.substr(pos);

    build_parser_state_machine(cap_str_value, cap->control_data_start_state);

    cap->cmds.emplace(cap_str_value->name,
                     cap_str_value);
}
